File: src/linkage_sim/core/mechanism.py
"""Mechanism: the top-level assembly of bodies, joints, and state.

A Mechanism owns the bodies, joints, and state vector mapping.
It provides the interface for building mechanisms and is the input
to all solvers.
"""

from __future__ import annotations

from dataclasses import dataclass, field

import numpy as np
from numpy.typing import NDArray

from typing import Callable

from linkage_sim.core.bodies import Body
from linkage_sim.core.constraints import (
    Constraint,
    FixedJoint,
    PrismaticJoint,
    RevoluteJoint,
    make_fixed_joint,
    make_prismatic_joint,
    make_revolute_joint,
)
from linkage_sim.core.drivers import (
    RevoluteDriver,
    constant_speed_driver,
    make_revolute_driver,
)
from linkage_sim.core.state import GROUND_ID, State


@dataclass
class Mechanism:
    """A planar mechanism: bodies connected by joint constraints.

    Usage:
        mech = Mechanism()
        mech.add_body(ground)
        mech.add_body(crank)
        mech.add_revolute_joint("J1", "ground", "O2", "crank", "A")
        mech.build()  # finalizes state vector, ready for solvers
    """

    _bodies: dict[str, Body] = field(default_factory=dict)
    _joints: list[Constraint] = field(default_factory=list)
    _state: State = field(default_factory=State)
    _built: bool = False

    @property
    def state(self) -> State:
        return self._state

    @property
    def bodies(self) -> dict[str, Body]:
        return self._bodies

    @property
    def joints(self) -> list[Constraint]:
        return self._joints

    @property
    def n_constraints(self) -> int:
        """Total number of constraint equations across all joints."""
        return sum(j.n_equations for j in self._joints)

    def add_body(self, body: Body) -> None:
        """Add a body to the mechanism.

        Must be called before build(). Ground and moving bodies are both added
        here; ground is recognized by its ID and excluded from the state vector.
        """
        if self._built:
            raise RuntimeError("Cannot add bodies after build().")
        if body.id in self._bodies:
            raise ValueError(f"Body '{body.id}' already exists in the mechanism.")
        self._bodies[body.id] = body

    def add_revolute_joint(
        self,
        joint_id: str,
        body_i_id: str,
        point_i_name: str,
        body_j_id: str,
        point_j_name: str,
    ) -> None:
        """Add a revolute joint between two bodies at named attachment points.

        Both bodies must already be added to the mechanism.
        """
        if self._built:
            raise RuntimeError("Cannot add joints after build().")

        body_i = self._get_body(body_i_id)
        body_j = self._get_body(body_j_id)
        pt_i = body_i.get_attachment_point(point_i_name)
        pt_j = body_j.get_attachment_point(point_j_name)

        joint = make_revolute_joint(
            joint_id=joint_id,
            body_i_id=body_i_id,
            point_i_name=point_i_name,
            point_i_local=pt_i,
            body_j_id=body_j_id,
            point_j_name=point_j_name,
            point_j_local=pt_j,
        )
        self._joints.append(joint)

    def add_fixed_joint(
        self,
        joint_id: str,
        body_i_id: str,
        point_i_name: str,
        body_j_id: str,
        point_j_name: str,
        delta_theta_0: float = 0.0,
    ) -> None:
        """Add a fixed joint that locks all relative motion."""
        if self._built:
            raise RuntimeError("Cannot add joints after build().")

        body_i = self._get_body(body_i_id)
        body_j = self._get_body(body_j_id)
        pt_i = body_i.get_attachment_point(point_i_name)
        pt_j = body_j.get_attachment_point(point_j_name)

        joint = make_fixed_joint(
            joint_id=joint_id,
            body_i_id=body_i_id,
            point_i_name=point_i_name,
            point_i_local=pt_i,
            body_j_id=body_j_id,
            point_j_name=point_j_name,
            point_j_local=pt_j,
            delta_theta_0=delta_theta_0,
        )
        self._joints.append(joint)

    def add_prismatic_joint(
        self,
        joint_id: str,
        body_i_id: str,
        point_i_name: str,
        body_j_id: str,
        point_j_name: str,
        axis_local_i: NDArray[np.float64],
        delta_theta_0: float = 0.0,
    ) -> None:
        """Add a prismatic joint that allows sliding along one axis.

        Body i owns the slide axis. Body j slides along it.
        Both bodies must already be added to the mechanism.

        Args:
            joint_id: Unique identifier for this joint.
            body_i_id: ID of the body that owns the slide axis.
            point_i_name: Name of the attachment point on body_i.
            body_j_id: ID of the sliding body.
            point_j_name: Name of the attachment point on body_j.
            axis_local_i: Unit vector along the slide axis in body_i's local frame.
            delta_theta_0: Initial relative angle θⱼ - θᵢ to maintain.
        """
        if self._built:
            raise RuntimeError("Cannot add joints after build().")

        body_i = self._get_body(body_i_id)
        body_j = self._get_body(body_j_id)
        pt_i = body_i.get_attachment_point(point_i_name)
        pt_j = body_j.get_attachment_point(point_j_name)

        joint = make_prismatic_joint(
            joint_id=joint_id,
            body_i_id=body_i_id,
            point_i_name=point_i_name,
            point_i_local=pt_i,
            body_j_id=body_j_id,
            point_j_name=point_j_name,
            point_j_local=pt_j,
            axis_local_i=axis_local_i,
            delta_theta_0=delta_theta_0,
        )
        self._joints.append(joint)

    def add_revolute_driver(
        self,
        driver_id: str,
        body_i_id: str,
        body_j_id: str,
        f: Callable[[float], float],
        f_dot: Callable[[float], float],
        f_ddot: Callable[[float], float],
    ) -> None:
        """Add a revolute driver prescribing relative angle vs. time.

        Adds one constraint equation: θⱼ - θᵢ - f(t) = 0.
        Both bodies must already be added to the mechanism.
        """
        if self._built:
            raise RuntimeError("Cannot add drivers after build().")

        self._get_body(body_i_id)
        self._get_body(body_j_id)

        driver = make_revolute_driver(
            driver_id=driver_id,
            body_i_id=body_i_id,
            body_j_id=body_j_id,
            f=f,
            f_dot=f_dot,
            f_ddot=f_ddot,
        )
        self._joints.append(driver)

    def add_constant_speed_driver(
        self,
        driver_id: str,
        body_i_id: str,
        body_j_id: str,
        omega: float,
        theta_0: float = 0.0,
    ) -> None:
        """Add a constant-speed revolute driver.

        f(t) = theta_0 + omega * t, f'(t) = omega, f''(t) = 0.
        """
        if self._built:
            raise RuntimeError("Cannot add drivers after build().")

        self._get_body(body_i_id)
        self._get_body(body_j_id)

        driver = constant_speed_driver(
            driver_id=driver_id,
            body_i_id=body_i_id,
            body_j_id=body_j_id,
            omega=omega,
            theta_0=theta_0,
        )
        self._joints.append(driver)

    def build(self) -> None:
        """Finalize the mechanism: register moving bodies in the state vector.

        After build(), no more bodies or joints can be added.
        The state vector size is fixed and solvers can operate.
        """
        if self._built:
            raise RuntimeError("Mechanism already built.")

        # Register all non-ground bodies in the state vector
        for body_id in sorted(self._bodies.keys()):
            if body_id != GROUND_ID:
                self._state.register_body(body_id)

        self._built = True

    def _get_body(self, body_id: str) -> Body:
        if body_id not in self._bodies:
            raise KeyError(
                f"Body '{body_id}' not found. "
                f"Available: {list(self._bodies.keys())}"
            )
        return self._bodies[body_id]

File: linkage-sim-rs/src/core/mechanism.rs
//! Mechanism: the top-level assembly of bodies, joints, and state.
//!
//! A Mechanism owns the bodies, joints, and state vector mapping.
//! It provides the interface for building mechanisms and is the input
//! to all solvers.

use nalgebra::Vector2;
use std::collections::HashMap;
use thiserror::Error;

use crate::core::body::Body;
use crate::core::constraint::{
    make_fixed_joint, make_prismatic_joint, make_revolute_joint, Constraint, JointConstraint,
};
use crate::core::driver::{constant_speed_driver, make_revolute_driver, RevoluteDriver};
use crate::core::state::{State, GROUND_ID};

/// A planar mechanism: bodies connected by joint constraints.
pub struct Mechanism {
    bodies: HashMap<String, Body>,
    joints: Vec<JointConstraint>,
    drivers: Vec<RevoluteDriver>,
    state: State,
    built: bool,
}

impl Mechanism {
    pub fn new() -> Self {
        Self {
            bodies: HashMap::new(),
            joints: Vec::new(),
            drivers: Vec::new(),
            state: State::new(),
            built: false,
        }
    }

    pub fn state(&self) -> &State {
        &self.state
    }

    pub fn bodies(&self) -> &HashMap<String, Body> {
        &self.bodies
    }

    pub fn joints(&self) -> &[JointConstraint] {
        &self.joints
    }

    pub fn n_drivers(&self) -> usize {
        self.drivers.len()
    }

    pub fn is_built(&self) -> bool {
        self.built
    }

    /// Total number of constraint equations across all joints and drivers.
    pub fn n_constraints(&self) -> usize {
        let joint_eqs: usize = self.joints.iter().map(|j| j.n_equations()).sum();
        let driver_eqs: usize = self.drivers.iter().map(|d| d.n_equations()).sum();
        joint_eqs + driver_eqs
    }

    /// Iterate over all constraints (joints + drivers) in order.
    pub fn all_constraints(&self) -> Vec<&dyn Constraint> {
        let mut all: Vec<&dyn Constraint> = Vec::new();
        for j in &self.joints {
            all.push(j);
        }
        for d in &self.drivers {
            all.push(d);
        }
        all
    }

    /// Add a body to the mechanism.
    pub fn add_body(&mut self, body: Body) -> Result<(), MechanismError> {
        if self.built {
            return Err(MechanismError::AlreadyBuilt("add body"));
        }
        if self.bodies.contains_key(&body.id) {
            return Err(MechanismError::DuplicateBody(body.id.clone()));
        }
        self.bodies.insert(body.id.clone(), body);
        Ok(())
    }

    /// Add a revolute joint between two bodies at named attachment points.
    pub fn add_revolute_joint(
        &mut self,
        joint_id: &str,
        body_i_id: &str,
        point_i_name: &str,
        body_j_id: &str,
        point_j_name: &str,
    ) -> Result<(), MechanismError> {
        if self.built {
            return Err(MechanismError::AlreadyBuilt("add joint"));
        }
        let pt_i = *self.get_attachment_point(body_i_id, point_i_name)?;
        let pt_j = *self.get_attachment_point(body_j_id, point_j_name)?;

        let joint = make_revolute_joint(joint_id, body_i_id, pt_i, body_j_id, pt_j);
        self.joints.push(JointConstraint::Revolute(joint));
        Ok(())
    }

    /// Add a fixed joint that locks all relative motion.
    pub fn add_fixed_joint(
        &mut self,
        joint_id: &str,
        body_i_id: &str,
        point_i_name: &str,
        body_j_id: &str,
        point_j_name: &str,
        delta_theta_0: f64,
    ) -> Result<(), MechanismError> {
        if self.built {
            return Err(MechanismError::AlreadyBuilt("add joint"));
        }
        let pt_i = *self.get_attachment_point(body_i_id, point_i_name)?;
        let pt_j = *self.get_attachment_point(body_j_id, point_j_name)?;

        let joint = make_fixed_joint(joint_id, body_i_id, pt_i, body_j_id, pt_j, delta_theta_0);
        self.joints.push(JointConstraint::Fixed(joint));
        Ok(())
    }

    /// Add a prismatic joint that allows sliding along one axis.
    pub fn add_prismatic_joint(
        &mut self,
        joint_id: &str,
        body_i_id: &str,
        point_i_name: &str,
        body_j_id: &str,
        point_j_name: &str,
        axis_local_i: Vector2<f64>,
        delta_theta_0: f64,
    ) -> Result<(), MechanismError> {
        if self.built {
            return Err(MechanismError::AlreadyBuilt("add joint"));
        }
        let pt_i = *self.get_attachment_point(body_i_id, point_i_name)?;
        let pt_j = *self.get_attachment_point(body_j_id, point_j_name)?;

        let joint = make_prismatic_joint(
            joint_id,
            body_i_id,
            pt_i,
            body_j_id,
            pt_j,
            axis_local_i,
            delta_theta_0,
        )
        .map_err(|e| MechanismError::InvalidJoint(e.to_string()))?;
        self.joints.push(JointConstraint::Prismatic(joint));
        Ok(())
    }

    /// Add a constant-speed revolute driver.
    pub fn add_constant_speed_driver(
        &mut self,
        driver_id: &str,
        body_i_id: &str,
        body_j_id: &str,
        omega: f64,
        theta_0: f64,
    ) -> Result<(), MechanismError> {
        if self.built {
            return Err(MechanismError::AlreadyBuilt("add driver"));
        }
        self.validate_body_exists(body_i_id)?;
        self.validate_body_exists(body_j_id)?;

        let driver = constant_speed_driver(driver_id, body_i_id, body_j_id, omega, theta_0);
        self.drivers.push(driver);
        Ok(())
    }

    /// Add a revolute driver with custom functions.
    pub fn add_revolute_driver(
        &mut self,
        driver_id: &str,
        body_i_id: &str,
        body_j_id: &str,
        f: impl Fn(f64) -> f64 + Send + Sync + 'static,
        f_dot: impl Fn(f64) -> f64 + Send + Sync + 'static,
        f_ddot: impl Fn(f64) -> f64 + Send + Sync + 'static,
    ) -> Result<(), MechanismError> {
        if self.built {
            return Err(MechanismError::AlreadyBuilt("add driver"));
        }
        self.validate_body_exists(body_i_id)?;
        self.validate_body_exists(body_j_id)?;

        let driver = make_revolute_driver(driver_id, body_i_id, body_j_id, f, f_dot, f_ddot);
        self.drivers.push(driver);
        Ok(())
    }

    /// Finalize the mechanism: register moving bodies in the state vector.
    pub fn build(&mut self) -> Result<(), MechanismError> {
        if self.built {
            return Err(MechanismError::AlreadyBuilt("build"));
        }

        let mut body_ids: Vec<_> = self.bodies.keys().cloned().collect();
        body_ids.sort();

        for body_id in &body_ids {
            if body_id != GROUND_ID {
                self.state
                    .register_body(body_id)
                    .map_err(|e| MechanismError::StateError(e.to_string()))?;
            }
        }

        self.built = true;
        Ok(())
    }

    fn get_attachment_point(
        &self,
        body_id: &str,
        point_name: &str,
    ) -> Result<&Vector2<f64>, MechanismError> {
        let body = self
            .bodies
            .get(body_id)
            .ok_or_else(|| MechanismError::BodyNotFound(body_id.to_string()))?;
        body.get_attachment_point(point_name)
            .map_err(|e| MechanismError::InvalidJoint(e.to_string()))
    }

    fn validate_body_exists(&self, body_id: &str) -> Result<(), MechanismError> {
        if !self.bodies.contains_key(body_id) {
            return Err(MechanismError::BodyNotFound(body_id.to_string()));
        }
        Ok(())
    }
}

impl Default for Mechanism {
    fn default() -> Self {
        Self::new()
    }
}

#[derive(Debug, Error)]
pub enum MechanismError {
    #[error("Cannot {0} after build()")]
    AlreadyBuilt(&'static str),
    #[error("Body '{0}' already exists in the mechanism")]
    DuplicateBody(String),
    #[error("Body '{0}' not found in mechanism")]
    BodyNotFound(String),
    #[error("Invalid joint: {0}")]
    InvalidJoint(String),
    #[error("State error: {0}")]
    StateError(String),
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::core::body::{make_bar, make_ground};
    use std::f64::consts::PI;

    fn build_fourbar() -> Mechanism {
        // Classic 4-bar: ground(O2)--crank--coupler--rocker--ground(O4)
        let ground = make_ground(&[("O2", 0.0, 0.0), ("O4", 0.038, 0.0)]);
        let crank = make_bar("crank", "A", "B", 0.01, 0.0, 0.0);
        let coupler = make_bar("coupler", "B", "C", 0.04, 0.0, 0.0);
        let rocker = make_bar("rocker", "C", "D", 0.03, 0.0, 0.0);

        let mut mech = Mechanism::new();
        mech.add_body(ground).unwrap();
        mech.add_body(crank).unwrap();
        mech.add_body(coupler).unwrap();
        mech.add_body(rocker).unwrap();

        mech.add_revolute_joint("J1", "ground", "O2", "crank", "A")
            .unwrap();
        mech.add_revolute_joint("J2", "crank", "B", "coupler", "B")
            .unwrap();
        mech.add_revolute_joint("J3", "coupler", "C", "rocker", "C")
            .unwrap();
        mech.add_revolute_joint("J4", "rocker", "D", "ground", "O4")
            .unwrap();
        mech.add_constant_speed_driver("D1", "ground", "crank", 2.0 * PI, 0.0)
            .unwrap();

        mech.build().unwrap();
        mech
    }

    #[test]
    fn fourbar_builds_correctly() {
        let mech = build_fourbar();
        assert!(mech.is_built());
        // 3 moving bodies × 3 coords each = 9
        assert_eq!(mech.state().n_coords(), 9);
        // 4 revolute joints × 2 + 1 driver = 9 constraints
        assert_eq!(mech.n_constraints(), 9);
    }

    #[test]
    fn cannot_add_after_build() {
        let mut mech = build_fourbar();
        let extra = make_bar("extra", "X", "Y", 0.1, 0.0, 0.0);
        assert!(mech.add_body(extra).is_err());
    }

    #[test]
    fn body_not_found_errors() {
        let mut mech = Mechanism::new();
        let ground = make_ground(&[("O", 0.0, 0.0)]);
        mech.add_body(ground).unwrap();
        assert!(mech
            .add_revolute_joint("J1", "ground", "O", "missing", "A")
            .is_err());
    }

    #[test]
    fn fourbar_constraint_count() {
        let mech = build_fourbar();
        assert_eq!(mech.n_constraints(), 9);
        assert_eq!(mech.all_constraints().len(), 5); // 4 joints + 1 driver
    }
}
